packages/runner/src/runner.ts

```diff
@@ -475,4 +474,346 @@ export class Runner {
       this.cancelListener();
     });
   }
+
+  async _injectWithdrawalRequest(
+    txSkeleton: TransactionSkeletonType,
+    packedl2Block: HexString
+  ): Promise<TransactionSkeletonType> {
+    const l2Block = new schemas.L2Block(
+      new Reader(packedl2Block).toArrayBuffer()
+    );
+    const rawL2Block = l2Block.getRaw();
+    const data: DataView = (rawL2Block as any).view;
+    const l2BlockHash = utils.ckbHash(data.buffer).serializeJson();
+    const l2BlockNumber =
+      "0x" + rawL2Block.getNumber().toLittleEndianBigUint64().toString(16);
+    const withdrawalRequestVec = l2Block.getWithdrawalRequests();
+    if (withdrawalRequestVec.length() === 0) {
+      return txSkeleton;
+    }
+    // add custodian lock dep
+    txSkeleton = txSkeleton.update("cellDeps", (cellDeps) => {
+      return cellDeps.push(this._deploymentConfig().custodian_lock_dep);
+    });
+    // collect all finalized and live custodian cells
+    let toBlockNumber =
+      BigInt(l2BlockNumber) -
+      BigInt(this._deploymentConfig().l2_finalized_period);
+    const toBlock = "0x" + toBlockNumber.toString(16);
+    const validCustodianCells = await this._queryValidCustodianCells(toBlock);
+    const deposition_block_hash = validCustodianCells[0].block_hash!;
+    const deposition_block_number = validCustodianCells[0].block_number!;
+    let ckbWithdrawalCapacity = 0n;
+    let sudtWithdrawalAssets: Map<HexString, BigInt> = new Map();
+    // build withdrawal cells
+    for (let i = 0; i < withdrawalRequestVec.length(); i++) {
+      const rawWithdrawalRequest = withdrawalRequestVec.indexAt(i).getRaw();
+      const withdrawalCapacity = new Reader(
+        rawWithdrawalRequest.getAmount().raw()
+      ).serializeJson();
+      // Record withdrawal assets info for inject input cells later
+      ckbWithdrawalCapacity =
+        ckbWithdrawalCapacity + BigInt(withdrawalCapacity);
+      const sudtScriptHash = new Reader(
+        rawWithdrawalRequest.getSudtScriptHash().raw()
+      ).serializeJson();
+      let withdrawalType: Script | undefined = undefined;
+      let outputData = "0x";
+      // check if it includes sudt withdrawal request
+      if (
+        sudtScriptHash !=
+        "0x0000000000000000000000000000000000000000000000000000000000000000"
```

Comment:
Why do we compare with a all zero script hash? I don't feel we know any script here whose hash is all zeros.

Reply:
In the case where user only withdraws CKB from L2, the `sudt_script_hash` would be simply set as all-zeros, and yes, it's not supposed to be found onchain. (Have confirmed it with @jjyr )